mmedit/models/backbones/sr_backbones/liif_net.py

```diff
@@ -0,0 +1,259 @@
+                areas.append(area + 1e-9)
+
+        tot_area = torch.stack(areas).sum(dim=0)
+        if self.local_ensemble:
+            t = areas[0]
+            areas[0] = areas[3]
+            areas[3] = t
+            t = areas[1]
+            areas[1] = areas[2]
+            areas[2] = t
+        result = 0
+        for pred, area in zip(preds, areas):
+            result = result + pred * (area / tot_area).unsqueeze(-1)
+
+        return result
+
+    def batched_predict(self, x, coord, cell):
+        """Batched predict.
+
+        Args:
+            x (Tensor): Input tensor.
+            coord (Tensor): coord tensor.
+            cell (Tensor): cell tensor.
+
+        Returns:
+            pred (Tensor): output of model.
+        """
+        with torch.no_grad():
+            n = coord.shape[1]
+            ql = 0
+            preds = []
+            while ql < n:
+                qr = min(ql + self.eval_bsize, n)
+                pred = self.query_rgb(x, coord[:, ql:qr, :], cell[:, ql:qr, :])
+                preds.append(pred)
+                ql = qr
+            pred = torch.cat(preds, dim=1)
+        return pred
+
+    def init_weights(self, pretrained=None, strict=True):
+        """Init weights for models.
+
+        Args:
+            pretrained (str, optional): Path for pretrained weights. If given
+                None, pretrained weights will not be loaded. Defaults to None.
+            strict (boo, optional): Whether strictly load the pretrained model.
+                Defaults to True.
+        """
+        if isinstance(pretrained, str):
+            logger = get_root_logger()
+            load_checkpoint(self, pretrained, strict=strict, logger=logger)
+        elif pretrained is not None:
+            raise TypeError('"pretrained" must be a str or None. '
+                            f'But received {type(pretrained)}.')
+
+
+@BACKBONES.register_module()
+class LIIFEDSR(LIIFNet):
+    """LIIF net based on EDSR.
+
+    Paper: Learning Continuous Image Representation with
+           Local Implicit Image Function
+
+    Args:
+        encoder (dict): Config for the generator.
+        imnet (dict): Config for the imnet.
+        local_ensemble (bool): Whether to use local ensemble. Default: True.
+        feat_unfold (bool): Whether to use feat unfold. Default: True.
+        cell_decode (bool): Whether to use cell decode. Default: True.
+        eval_bsize (int): Size of batched predict. Default: None.
+    """
+
+    def __init__(self,
+                 encoder,
+                 imnet,
+                 local_ensemble=True,
+                 feat_unfold=True,
+                 cell_decode=True,
+                 eval_bsize=None):
+        super().__init__(
+            encoder=encoder,
+            imnet=imnet,
+            local_ensemble=local_ensemble,
+            feat_unfold=feat_unfold,
+            cell_decode=cell_decode,
+            eval_bsize=eval_bsize)
+
+        self.conv_first = self.encoder.conv_first
```

Comment:
I think it is better to directly construct the EDSR modules here? In the current design, when someone accidentally uses other networks which do not have `conv_first`, then `self.encoder.conv_first` would lead to an error.

Reply:
This is LIIFEDSR, the backbone should be EDSR.

Reply:
But the encoder is not necessarily EDSR. It may not be the best way to define EDSR here.

Reply:
There is a chance that the user (accidentally) uses other backbones in LIIFEDSR. Maybe add a value error here?

Reply:
He should use other backbones.

Reply:
If the user uses SRCNN as `encoder` in LIIFEDSR, there will be an error since SRCNN has no `conv_first. 

Of course, this is not allowed. Therefore, it is better to raise an error so that the user could know why does the error occur. We can raise an ValueError when the EDSR is not used. 

Similar for line 43.

Reply:
Of course, if the user uses SRCNN as an `encoder` in LIIFEDSR, there will be an error. But we do not need to raise an error, because `python` will raise the following error itselfs 

`AttributeError: LIIF: LIIFEDSR: 'SRCNN' object has no attribute 'mid_channels'`

or

`AttributeError: LIIF: LIIFEDSR: 'SRCNN' object has no attribute 'conv_first'`

Why do we repeat the errors?

Reply:
They don't know the main reason is "SRCNN should not be used". I think for some users, they may not be able to relate "SRCNN has no mid_channels" and " We do not accept SRCNN here".

Even if no error is raised, we should at least tell the users in the docstring that only EDSR and RDN can be used. (LIIFEDSR may be fine as EDSR is in the class name, but LIIFNet should state this fact).

Reply:
LIIFNet is the base class of LIIF Net, and in LIIFNet (liif_net.py, line 18~19):

    The subclasses should define `generator` with `encoder` and `imnet`,
        and overwrite the function `gen_feature`.

LIIFEDSR is a subclass of LIIF, and LIIFRDN is in PR #428 .

User can add LIIFSRCNN or others.

Reply:
Then LIIFSRCNN cannot inherit `__init__` from LIIFNet, as there is no `mid_channels`?

Reply:
In this case, why don’t we move `build_backbone` to the subclasses?

Reply:
Calculate imnet_in_dim is a common process, which should be placed in `LIIFNet`.

 `imnet_in_dim = self.encoder.mid_channels`

So we need to build the backbone in `LIIFNet`.

1st, seldom backbone does have `mid_channels`
2ed, if the user wants to use these backbones like SRCNN, they can init the subclass without inheriting `__init__` from LIIFNet.